feat(components): add diary page

SukhanovPavel opened the pull request.

SukhanovPavel committed:
feat(components): add diary page

SukhanovPavel committed:
fix(components): fixed cretical errors

Commit:
fix(components): changed types for student page table

META DATA: issue44

Hellomyguest commented on src/features/data-table/DataTable.tsx:
```diff
@@ -0,0 +1,28 @@
+  form: FormInstance;
+  data: object[] | undefined;
+};
+export const DataTable = ({ form, data, columns }: Props) => (
```
Данный компонент-таблица используется только для дневника, предполагаю что название должно быть DiaryTable

Hellomyguest commented on src/features/data-table/DataTable.tsx:
```diff
@@ -0,0 +1,28 @@
+import { Form, FormInstance, Table } from "antd";
+import { DiaryEditableCell } from "../diary-editable-cell/DiaryEditableCell";
```
внутри папки с компонентом создай папку ui и перенеси компонент ячейки туда - так как это составляющая конкретно данной фичи

Hellomyguest commented on src/pages/sign/signUp/model.ts:
```diff
@@ -7,39 +7,41 @@ export const $users = createStore<User[]>([
+          id: 4,
+          title: "ДЗ 4",
+          deadline: "02.02.2023",
+          status: HomeworksStatus.REJECTED,
+        },
+        homework5: {
+          id: 5,
+          title: "ДЗ 5",
+          deadline: "01.03.2023",
+          status: HomeworksStatus.DEFAULT,
+        },
       },
-      homework2: {
-        id: 2,
-        title: "ДЗ 2",
-        deadline: "02.01.2023",
-        status: HomeworksStatus.APPROVED,
-      },
-      homework3: {
-        id: 3,
-        title: "ДЗ 3",
-        deadline: "01.02.2023",
-        status: HomeworksStatus.PENDING,
-      },
-      homework4: {
-        id: 4,
-        title: "ДЗ 4",
-        deadline: "02.02.2023",
-        status: HomeworksStatus.REJECTED,
-      },
-      homework5: {
-        id: 5,
-        title: "ДЗ 5",
-        deadline: "01.03.2023",
-        status: HomeworksStatus.DEFAULT,
-      },
-    },
+    ],
```
Предполагаю что Сергей имел ввиду что homeworks будет массивом вида:

`homeworks: [{id: "homework1", title: "ДЗ 1", deadline: "...", status: HomeworksStatus.DEFAULT}, {id: "homework2", title: "ДЗ 2", deadline: "...", status: HomeworksStatus.REJECTED}]`

в ином случае нам нет смысла оборачивать объект в массив

Hellomyguest commented on src/pages/sign/signUp/model.ts:
```diff
@@ -60,4 +63,5 @@ $users
     return state;
-  });
+  })
+  .on(saveHomework, (_, payload) => payload);
```
ты сохраняешь изменения по каждому пользователю отдельно, нет смысла передавать целиком массив всех пользователей, у нас есть event updateUser - просто передавай в него обновленный объект конкретного пользователя

Hellomyguest commented on src/pages/diary/model.ts:
```diff
@@ -0,0 +1,28 @@
+export const $studentDiary = $users.map((state) =>
+  state.map((user) => {
+    const userHomeworks = Array.isArray(user.homeworks)
+      ? user.homeworks.reduce(
+          (accum, item: string | { [key: string]: string } | any) => ({
+            ...accum,
+            key: item.id,
+            [item.key]: { item },
+          })
+        )
+      : {};
+
+    return {
+      id: user.id,
+      fullName: `${user.name} ${user.surname}`,
+      ...userHomeworks,
+    };
```
user.homeworks это массив из одного объекта, считаю что reduce тут ни к чему, при этом у каждого студента будет ключ "homeworks" - а у каждой строки должен быть индивидуальный ключ. 
key нам здесь по сути вообще не нужен в andt компонент Table можно передать пропс rowKey="id" и в качестве key строки будет использоваться id

```
return {
      id: user.id,
      fullName: `${user.name} ${user.surname}`,
      ...user.homeworks[0],
    };
```

Hellomyguest commented on src/pages/diary/page.tsx:
```diff
@@ -1,5 +1,237 @@
+  const [form] = Form.useForm();
+  const isDiaryEditing = (store: User) => store.id === editingDiaryKey;
+
+  const editDiary = (store: Partial<User> & { key?: string }) => {
+    form.setFieldsValue({
+      homeworks: [
+        {
+          homework1: { status: "" },
+          homework2: { status: "" },
+          homework3: { status: "" },
+          homework4: { status: "" },
+          homework5: { status: "" },
+        },
+      ],
+      ...store,
+    });
+    setEditingDiaryKeyFn(typeof store.id === "string" ? store.id : "");
+  };
+
+  const cancelDiaryEdit = () => {
+    setEditingDiaryKeyFn("");
+  };
```
Не знаю насколько нам нужны эти кнопки редактировать, сохранить, отменить. 
я бы предпочел использовать вариант таблицы где все ячейки редактируются без лишних действий: 
[https://ant.design/components/table#components-table-demo-edit-cell](url)
единственный недочёт данного подхода, в сравнении с используемым, это потеря в производительности, так как на каждое изменение идёт обновление стора пользователей, но так как на потоке не так уж и много студентов думаю что это не критично.
И если прям рассуждать то вряд ли в момент будет изменяться оценка более чем по одному ДЗ студента, так что при текущей реализации просто придётся каждый раз нажимать 2 лишних кнопки

В общем это надо обсуждать с Сергеем )

Hellomyguest commented on src/pages/diary/page.tsx:
```diff
@@ -1,5 +1,237 @@
+      const row = (await form.validateFields()) as Homeworks;
+
+      const newData = [...users];
+      const index = newData.findIndex((item) => key === item.id);
+      if (index > -1) {
+        const editingUser = newData[index];
```
просто find:

`const editingUser = newData.find((item) => key === item.id);`

Hellomyguest commented on src/pages/diary/page.tsx:
```diff
@@ -1,5 +1,237 @@
+      const index = newData.findIndex((item) => key === item.id);
+      if (index > -1) {
+        const editingUser = newData[index];
+        newData.splice(index, 1, {
+          ...editingUser,
+          homeworks: [row],
+        });
+        saveHomeworkFn(newData);
```
1. row это объект формата {[колонка]: {status: [значение ячейки]}}, таким образом затираются все остальные ключи каждого дз( id, deadline, title), поэтому нужно сначала отдельно вычислить новый объект homeworks пользователя, примерно так(assignIn - из билиотеки Lodash):

` const updatedHomeworks = editingUser.homeworks ? Object.keys(editingUser.homeworks).reduce((homeworks, homework) => ({...homeworks, [homework]: assignIn(editingUser.homeworks[homework],row[homework])}), {}) : {};
const updatedUser = {...editingUser, homeworks: updatedHomeworks}
`

2. не нужно изменять массив всех пользователей и потом его передавать, у нас есть event updateUser в который просто передаётся обновлённый объект пользователя, сам event уже отредактирует стор.

`updateUser(updatedUser)`

Hellomyguest commented on src/pages/diary/model.ts:
```diff
@@ -0,0 +1,28 @@
+  state.map((user) => {
+    const userHomeworks = Array.isArray(user.homeworks)
+      ? user.homeworks.reduce(
+          (accum, item: string | { [key: string]: string } | any) => ({
```
не используем any

SukhanovPavel committed:
fix(components): fix user.homeworks store

SukhanovPavel committed:
feat: add editable cell in diary page

SukhanovPavel committed:
fix(components): fix click outside

Hellomyguest commented on src/features/diary-table/DiaryTable.tsx:
```diff
@@ -0,0 +1,238 @@
+
+  if (editable) {
+    childNode = editing ? (
+      <Form.Item
+        style={{ margin: 0 }}
+        name={dataIndex}
+        initialValue={HomeworksStatus.default}
+        rules={[
+          {
+            required: true,
+            message: `${title} is required.`,
+          },
+        ]}
+      >
+        <Select
+          options={[
+            { value: "approved", label: "Зачет" },
+            { value: "pending", label: "Выполняется" },
+            { value: "rejected", label: "Незачет" },
+            { value: "default", label: "Не выбрано" },
+          ]}
+          ref={inputRef}
+          onChange={save}
+          // onPressEnter={save}
+          onBlur={cansel}
+        />
+      </Form.Item>
+    ) : (
+      <div
+        className="editable-cell-value-wrap"
+        style={{ paddingRight: 24 }}
+        onClick={toggleEdit}
+      >
+        {children}
+      </div>
+    );
+  }
+
+  return <td {...restProps}>{childNode}</td>;
+};
```
Вынести в отдельный компонент

Hellomyguest commented on src/features/diary-table/DiaryTable.tsx:
```diff
@@ -0,0 +1,238 @@
+
+type ColumnTypes = Exclude<EditableTableProps["columns"], undefined>;
+
+export const DiaryTable: React.FC = () => {
```
нет необходимости в React.FC

Hellomyguest commented on src/features/diary-table/DiaryTable.tsx:
```diff
@@ -0,0 +1,238 @@
+      dataIndex: "fullName",
+      editable: false,
+      width: "15%",
+    },
+    {
+      title: "ДЗ-1",
+      dataIndex: ["homework1", "status"],
+      width: "15%",
+      editable: true,
+      render: (status: string) => HomeworksStatus[status],
+    },
+    {
+      title: "ДЗ-2",
+      dataIndex: ["homework2", "status"],
+      width: "15%",
+      editable: true,
+      render: (status: string) => HomeworksStatus[status],
+    },
+    {
+      title: "ДЗ-3",
+      dataIndex: ["homework3", "status"],
+      width: "15%",
+      editable: true,
+      render: (status: string) => HomeworksStatus[status],
+    },
+    {
+      title: "ДЗ-4",
+      dataIndex: ["homework4", "status"],
+      width: "15%",
+      editable: true,
+      render: (status: string) => HomeworksStatus[status],
+    },
+    {
+      title: "ДЗ-5",
+      dataIndex: ["homework5", "status"],
+      width: "15%",
+      editable: true,
+      render: (status: string) => HomeworksStatus[status],
+    },
+  ];
```
в pages/student/model есть стор колонок домашних заданий который собирается из расписания.

Hellomyguest commented on src/features/diary-table/DiaryTable.tsx:
```diff
@@ -0,0 +1,238 @@
+  index: number;
+}
+
+const EditableRow: React.FC<EditableRowProps> = ({ index, ...props }) => {
+  const [form] = Form.useForm();
+  return (
+    <Form form={form} component={false}>
+      <EditableContext.Provider value={form}>
+        <tr {...props} />
+      </EditableContext.Provider>
+    </Form>
+  );
+};
```
Вынести в отдельный компонент

Hellomyguest commented on src/pages/sign/signUp/model.ts:
```diff
@@ -17,6 +17,38 @@ export const $users = createStore<User[]>([
     nickName: "admin",
     name: "Сергей",
     surname: "Махнаткин",
+    homeworks: [
+      {
+        id: "homework1",
+        title: "ДЗ 1",
+        deadline: "01.01.2023",
+        status: HomeworksStatus.APPROVED,
+      },
+      {
+        id: "homework2",
+        title: "ДЗ 2",
+        deadline: "02.01.2023",
+        status: HomeworksStatus.APPROVED,
+      },
+      {
+        id: "homework3",
+        title: "ДЗ 3",
+        deadline: "01.02.2023",
+        status: HomeworksStatus.PENDING,
+      },
+      {
+        id: "homework4",
+        title: "ДЗ 4",
+        deadline: "02.02.2023",
+        status: HomeworksStatus.REJECTED,
+      },
+      {
+        id: "homework5",
+        title: "ДЗ 5",
+        deadline: "01.03.2023",
+        status: HomeworksStatus.DEFAULT,
+      },
+    ],
```
моки не нужны, домашние задания строятся автоматом на основе расписания

Hellomyguest commented on src/features/diary-table/DiaryTable.tsx:
```diff
@@ -0,0 +1,238 @@
+
+  let childNode = children;
+
+  enum HomeworksStatus {
+    default = "Не выбрано",
+    approved = "Зачет",
+    pending = "Выполняется",
+    rejected = "Незачет",
+  }
```
pages/sign/signup/lib/types есть этот enum

Hellomyguest commented on src/features/diary-table/DiaryTable.tsx:
```diff
@@ -0,0 +1,238 @@
+      <Form.Item
+        style={{ margin: 0 }}
+        name={dataIndex}
+        initialValue={HomeworksStatus.default}
```
думаю initialValue ни к чему

Commit:
refactor(components): diary table edited

META DATA: issue44

Commit:
feat(components): add users sorting

META DATA: issue44

Commit:
fix(components): changed fullname in diary table

META DATA: issue44

dpolevodin commented on src/features/diary-table/ui/DiaryEditableCell/DiaryEditableCell.tsx:
```diff
@@ -0,0 +1,120 @@
+        selectRef.current.focus();
+      }
+    }
+  }, [editing]);
```
Мне кажется, не хватает selectRef в массиве зависимостей, возможно эффект сейчас работает не совсем так, как задумывалось

Hellomyguest replied:
Перепроверил, всё работает корректно(useEffect срабатывает, в момент начала редактирования элемента.), добавление selectRef в массив зависимостей не оказывает никакого эффекта.

dpolevodin commented on src/features/diary-table/ui/DiaryEditableCell/DiaryEditableCell.tsx:
```diff
@@ -0,0 +1,120 @@
+    }
+  };
+
+  const cansel = () => setEditing(!editing);
```
Возможно здесь ошибка в названии функции

Hellomyguest replied:
исправил

dpolevodin commented on src/features/diary-table/ui/DiaryEditableCell/DiaryEditableCell.tsx:
```diff
@@ -0,0 +1,120 @@
+
+  const cansel = () => setEditing(!editing);
+
+  let childNode = children;
```
Не совсем понятна целесообразность объявления переменной? Почему бы просто не использовать children?

Hellomyguest replied:
Это сделано, так как нельзя обращаться к параметру функции напрямую(правило eslint)

dpolevodin commented on src/pages/diary/model.ts:
```diff
@@ -0,0 +1,22 @@
+import { $users } from "../sign/signUp/model";
+import { UserHomeworkType } from "../sign/signUp/lib/types";
+
+export const $studentDiary = $users.map((state) =>
```
Не хватает типа для стора

Hellomyguest replied:
Добавил типизацию

Commit:
fix(components): fix naming and typing

changed name from cansel to cancel in diaryeditablecell
added type for \\$studentDiary store

META DATA: issue44
